File: go-probe/watchdog.go
// Сторож связи с сервером — аналог C# ServerWatchdogService.
//
// Если сервер не обращался к пробе дольше Probe:ServerTimeoutHours часов
// (0 — сторож выключен), проба считает себя удалённой: останавливает все задачи
// по расписанию, удаляет реестр (TaskInfo.json) и очищает кэш недоставленных
// результатов (JobResult.json). HTTP продолжает слушаться: если сервер вернётся,
// его фоновая сверка восстановит задачи автоматически.
package main

import (
	"context"
	"sync/atomic"
	"time"
)

// watchdogCheckInterval — период проверки сторожа.
const watchdogCheckInterval = 5 * time.Minute

// ContactTracker — потокобезопасная отметка последнего обращения сервера.
type ContactTracker struct {
	lastContact atomic.Int64 // UnixNano
}

// NewContactTracker создаёт трекер с отсчётом от запуска пробы.
func NewContactTracker() *ContactTracker {
	t := &ContactTracker{}
	t.Mark()
	return t
}

// Mark фиксирует обращение сервера (вызывается на каждый api-запрос).
func (t *ContactTracker) Mark() { t.lastContact.Store(time.Now().UnixNano()) }

// Last возвращает момент последнего обращения сервера.
func (t *ContactTracker) Last() time.Time { return time.Unix(0, t.lastContact.Load()) }

// RunWatchdog — цикл сторожа: молчание сервера дольше timeoutHours приводит
// к остановке всех задач и очистке кэша (однократно до следующего контакта).
func RunWatchdog(ctx context.Context, timeoutHours int, tracker *ContactTracker,
	tasks *TaskRegistry, results *ResultStore) {

	if timeoutHours <= 0 {
		logWatchdog.Info("Сторож связи выключен (Probe:ServerTimeoutHours = 0)")
		return
	}
	logWatchdog.Info("Сторож связи запущен", "порог_молчания_ч", timeoutHours)

	timeout := time.Duration(timeoutHours) * time.Hour
	var lastCleared time.Time

	ticker := time.NewTicker(watchdogCheckInterval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			last := tracker.Last()
			// Уже чистили после этого контакта — ждём следующего обращения сервера.
			if !lastCleared.Before(last) || time.Since(last) < timeout {
				continue
			}

			stopped := tasks.ClearAll()
			results.Clear()
			lastCleared = time.Now()
			logWatchdog.Warn("Сервер молчит дольше порога — проба считает себя удалённой, всё очищено",
				"последний_контакт", last.Format("02.01.2006 15:04"),
				"порог_ч", timeoutHours, "остановлено_задач", stopped)
		}
	}
}

File: rust-probe/src/watchdog.rs
//! Сторож связи с сервером — аналог Go RunWatchdog / C# ServerWatchdogService.
//!
//! Если сервер не обращается к пробе дольше Probe:ServerTimeoutHours, проба считает,
//! что её удалили: останавливает все задачи по расписанию и чистит кэш результатов.
//! Любое обращение сервера сбрасывает отсчёт.

use crate::logging::{f, Logger};
use crate::registry::TaskRegistry;
use crate::resultstore::ResultStore;
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::Arc;
use std::time::Duration;

static LOG: Logger = Logger("watchdog");

/// Как часто проверять молчание сервера.
const CHECK_INTERVAL: Duration = Duration::from_secs(60);

/// Отметчик последнего обращения сервера (обновляется HTTP-обработчиками).
pub struct ContactTracker {
    last: AtomicI64, // Unix-время в секундах
}

impl ContactTracker {
    pub fn new() -> Self {
        ContactTracker { last: AtomicI64::new(now_unix()) }
    }

    /// Фиксирует обращение сервера.
    pub fn mark(&self) {
        self.last.store(now_unix(), Ordering::Relaxed);
    }

    /// Время последнего обращения (Unix-секунды).
    pub fn last(&self) -> i64 {
        self.last.load(Ordering::Relaxed)
    }
}

/// Текущее время в секундах Unix.
fn now_unix() -> i64 {
    chrono::Local::now().timestamp()
}

/// Запускает сторож связи в отдельном потоке.
pub fn start(timeout_hours: i64, tracker: Arc<ContactTracker>, tasks: Arc<TaskRegistry>, results: Arc<ResultStore>) {
    if timeout_hours <= 0 {
        LOG.info("Сторож связи выключен (Probe:ServerTimeoutHours = 0)", &[]);
        return;
    }
    LOG.info("Сторож связи запущен", &[f("порог_молчания_ч", timeout_hours)]);

    std::thread::spawn(move || {
        let timeout_secs = timeout_hours * 3600;
        let mut last_cleared: i64 = 0;
        loop {
            std::thread::sleep(CHECK_INTERVAL);

            let last = tracker.last();
            // Уже чистили после этого контакта — ждём следующего обращения сервера.
            if last_cleared >= last || now_unix() - last < timeout_secs {
                continue;
            }

            let stopped = tasks.clear_all();
            results.clear();
            last_cleared = now_unix();
            let when = chrono::DateTime::from_timestamp(last, 0)
                .map(|t| t.with_timezone(&chrono::Local).format("%d.%m.%Y %H:%M").to_string())
                .unwrap_or_default();
            LOG.warn(
                "Сервер молчит дольше порога — проба считает себя удалённой, всё очищено",
                &[f("последний_контакт", when), f("порог_ч", timeout_hours), f("остановлено_задач", stopped)],
            );
        }
    });
}
